File: RB101/small_problems/medium_1.rb
=begin
# Rotation (Part 1)
def rotate_array(arr)
  rotated_arr = arr.slice(1, arr.size - 1)
  rotated_arr << arr[0]
end

p rotate_array([7, 3, 5, 2, 9, 1]) == [3, 5, 2, 9, 1, 7]
p rotate_array(['a', 'b', 'c']) == ['b', 'c', 'a']
p rotate_array(['a']) == ['a']

x = [1, 2, 3, 4]
p rotate_array(x) == [2, 3, 4, 1]   # => true
p x == [1, 2, 3, 4]                 # => true

# Rotation (Part 2)
def rotate_rightmost_digits(num, digits)
  num_arr = num.to_s.chars 
  slice_index = num_arr.length - digits
  rotate_nums = num_arr.slice!(slice_index, digits)
  new_arr = num_arr.concat(rotate_array(rotate_nums))
  new_arr.join.to_i
end


p rotate_rightmost_digits(735291, 1) == 735291
p rotate_rightmost_digits(735291, 2) == 735219
p rotate_rightmost_digits(735291, 3) == 735912
p rotate_rightmost_digits(735291, 4) == 732915
p rotate_rightmost_digits(735291, 5) == 752913
p rotate_rightmost_digits(735291, 6) == 352917

# Rotation (part 3)
def max_rotation(num)
  n = num.to_s.length
  until n == 1  
    num = rotate_rightmost_digits(num, n)
    n -= 1
  end
  num
end

p max_rotation(735291) == 321579
p max_rotation(3) == 3
p max_rotation(35) == 53
p max_rotation(105) == 15 # the leading zero gets dropped
p max_rotation(8_703_529_146) == 7_321_609_845

def initialize_switches(n)
  keys = (1..n).to_a
  keys.each_with_object({}){|key, hsh| hsh[key] = false}
end
# turn switch

def switches_to_turn(round, n)
  ending_mutiplier = n / round
  (1..ending_mutiplier).to_a.map{|num| num * round}
end


def turn_some_switches(n)
  switch_hsh = initialize_switches(n)
  round = 1
  switches_to_turn = {}
  loop do 
    switches_to_turn = switches_to_turn(round, n)
    switches_to_turn.each do |switch|
      switch_hsh[switch] = !switch_hsh[switch]
    end
    break if round == n
    round += 1
  end
  on = switch_hsh.select{|k, v| v == true}.keys
  off = switch_hsh.select{|k, v| v == false}.keys
  "lights #{on[0..-2].join(', ')} and #{on[-1]} are on, lights #{off[0..-2].join(', ')} and #{off[-1]} are off"
end

p turn_some_switches(5)

# diamonds
# iterate through an array of 1..n to print n number of lines
# within each line , the number of * to print +2 until == 9. then -2
# number of spaces to prinr = (n - number of * )/2

def half_spaces_array(n)
  ascending_spaces = (1..(n/2)).to_a
  ascending_spaces.reverse + [0] + ascending_spaces
end

def diamonds(n)
  half_spaces_array = half_spaces_array(n)
  half_spaces_array.each do |half_space|
    if half_space == n / 2
      puts " " * half_space + "*" + " " * half_space
    else
      space_between_diamonds = n - half_space * 2 - 2
      puts " " * half_space + "*" + " " * space_between_diamonds + "*" + " " * half_space
    end
  end
end

diamonds(11)

def minilang(str)
  stack = []
  register = 0
  str.split.each do |input|
    case input
    when 'PUSH'  then stack << register
    when 'ADD'   then register = register + stack.pop
    when 'SUB'   then register = register - stack.pop
    when 'MULT'  then register = register * stack.pop
    when 'DIV'   then register = register / stack.pop
    when 'MOD'   then register = register % stack.pop
    when 'POP'   then register = stack.pop
    when 'PRINT' then puts register
    when 'S'     then puts stack
    else         register = input.to_i
    end
  end
end

# minilang('PRINT')
# minilang('5 PUSH 3 MULT PRINT')
# minilang('5 PRINT PUSH 3 PRINT ADD PRINT')
# minilang('5 PUSH POP PRINT')
# minilang('3 PUSH 4 PUSH 5 PUSH PRINT ADD PRINT POP PRINT ADD PRINT')
# minilang('3 PUSH PUSH 7 DIV MULT PRINT ')
# minilang('4 PUSH PUSH 7 MOD MULT PRINT ')
# minilang('-3 PUSH 5 SUB PRINT')
# minilang('6 PUSH')
# (nothing printed; no PRINT commands)
minilang('3 PUSH 5 DIV PUSH S 7 PUSH 4 PUSH 5 MULT PRINT PUSH 3 ADD PRINT SUB PRINT DIV PRINT')

# Word to Digit

NUMBERS = {'one'=> '1', 'two'=> '2', 'three'=> '3', 'four'=> '4', 'five'=> '5', 'six'=>'6', 'seven'=> '7', 'eight'=> '8', 'nine'=> '9', 'zero'=> '0'}
def word_to_digit(str)
  NUMBERS.each_key do |word|
    str.gsub!(/\b#{word}\b/, NUMBERS[word])
  end
  trimed_string_arr = trim_numbers(str).split
  trimed_string_arr.map! do |word|
    format_phone_number(word)
  end

  trimed_string_arr.join(" ")

end


def trim_numbers(str)
  arr = str.chars
  numbers = NUMBERS.values
  n = 0
  loop do
    if numbers.include?(arr[n]) && numbers.include?(arr[n + 2]) && arr[n + 1] = ' '
      arr[n + 1] = ''
    end
    break if n == arr.size - 3
    n += 1
  end
  arr.join
end

def format_phone_number(word)
  letters = word.chars
  letters_trimed = word.gsub(/[^0-9a-z]/i, '').chars
  if letters_trimed.all?{|letter| NUMBERS.values.include?(letter)} && letters_trimed.size == 10
    letters.insert(0, "(")
    letters.insert(4, ") ")
    letters.insert(9, "-")
    letters.join("")
  else
    word
  end
  
end

p word_to_digit('Please call me at five five five one two three four four four four. freight Thanks.') 
# == 'Please call me at 5 5 5 1 2 3 4. Thanks.'



def fibonacci(n)
  if n <= 2
    1
  else
    fibonacci(n - 1) + fibonacci(n - 2)
  end
end

p fibonacci(5)

p fibonacci(1) == 1
p fibonacci(2) == 1
p fibonacci(3) == 2
p fibonacci(4) == 3
p fibonacci(5) == 5
p fibonacci(12) == 144
p fibonacci(20) == 6765

def fibonacci(n)
  n1 = 1
  n2 = 0
  f = 0
  counter = 0
  loop do 
    f = n1 + n2
    n1 = n2
    n2 = f
    counter += 1
    break if counter == n
  end
  f
end

p fibonacci(20) == 6765
p fibonacci(100) == 354224848179261915075
p fibonacci(100_001) # => 4202692702.....8285979669707537501
=end

def fibonacci_last(num)
  first, second = [1, 1]
  3.upto(num) do
    first, second = [second, first + second]
  end
  second.to_s[-1].to_i
end

p fibonacci_last(100)

File: JS215/exercises/medium1.js
// function rotateArray(arr) {
//   let newArr
//   if (!Array.isArray(arr)) {
//     return undefined;
//   } else {
//     if (arr.length === 0) return [];
//     newArr = arr.slice();
//     let firstElement = newArr.shift();
//     newArr.push(firstElement)
//   }
//   return newArr;
// }

// console.log(rotateArray([7, 3, 5, 2, 9, 1]));       // [3, 5, 2, 9, 1, 7]
// console.log(rotateArray(['a', 'b', 'c']));          // ["b", "c", "a"]
// console.log(rotateArray(['a']));                    // ["a"]
// console.log(rotateArray([1, 'a', 3, 'c']));         // ["a", 3, "c", 1]
// console.log(rotateArray([{ a: 2 }, [1, 2], 3]));    // [[1, 2], 3, { a: 2 }]
// console.log(rotateArray([]));                       // []

// // return `undefined` if the argument is not an array
// console.log(rotateArray());                         // undefined
// console.log(rotateArray(1));                        // undefined


// // the input array is not mutated
// const array = [1, 2, 3, 4];
// console.log(rotateArray(array));                    // [2, 3, 4, 1]
// array;                                 // [1, 2, 3, 4]

// P roatate digits in a number. 
    // input the number to be rotated and the number of digits from the left to be roatetd
    // output is the rotated number
// E: when 1, no need to do anything
     // more than 1. count the nth digit from the left and move it to the end of the number
// D: string => array
// A: 1. number to string. string to array
  //  2. use splice(n, 1) to remove the element from the array and return the element. save that element to a variable
  // 3. concat that element to the end of the original array. and join and convert to number

// function rotateRightmostDigits(num, n) {
//   let arr = num.toString().split('');
//   let temp = arr.splice(-n, 1);
//   return Number(arr.concat(temp).join(''));
// }

// console.log(rotateRightmostDigits(735291, 1));      // 735291
// console.log(rotateRightmostDigits(735291, 2));      // 735219
// console.log(rotateRightmostDigits(735291, 3));      // 735912
// console.log(rotateRightmostDigits(735291, 4));      // 732915
// console.log(rotateRightmostDigits(735291, 5));      // 752913
// console.log(rotateRightmostDigits(735291, 6));      // 352917

// P: input is a number Ouput is a rotated number
// E: 105
    // 105 - 051 i = 0
    // 051 - 015 i = 1
// D string
// A: iterate through each position in the string and perform rotation at each iteration
    // take the part of the string from the iteration index position to the end of the string
    // for that substring, move the first element of the string to the last element
    // return the first half plus the mutated substring
    // go to the next iteration
// return the final string and convert to number

// function maxRotation(num) {
//   let numStr = num.toString();
//   for (i = 0; i < numStr.length; i += 1) {
//     let secondHalf = numStr.slice(i);
//     let firstHalf = numStr.slice(0, i);
//     secondHalf = rotate(secondHalf);
//     numStr = firstHalf + secondHalf;
//   }
//   return Number(numStr);
// }

// function rotate(str) {
//   return str.slice(1) + str[0];
// }

// console.log(maxRotation(735291));          // 321579
// console.log(maxRotation(3));               // 3
// console.log(maxRotation(35));              // 53
// console.log(maxRotation(105));             // 15 -- the leading zero gets dropped
// console.log(maxRotation(8703529146));      // 7321609845

//P input: a string of numbers and actions
  // ouput: nothing unless print action is called if print is called print the regiter value to the console

//E minilang('3 PUSH PUSH 7 DIV MULT PRINT');
  // 3: place 2 in the register
  // push: [3] in the stack and 3 is still in the register;
  // push: [3, 3] in the stack and 3 is still in the register
  // 7: [3, 3] 7 register
  // DIV: [3] 7 / 3 = 2, 2 is in the register
  // MULT: [] 3 * 2 = 6, 6 is in the register
  // print,print 6 to the register
  // other then numbers and print, every function modify the stack, 
    // push adds to the stack, do not modify the register;
    // math opertions pop from the stack, perform some operation update register with the return value
//array 
//A:
  // start with an empty array as stack
  // and a 0 as register value
  // helper functions for each operation
  // split the program string into array
  // iterate through the array and use switch statement to match the command with the right function and then go to the next iteration;

// function minilang(str) {
//   let stack = [];
//   let register = 0;
//   let programs = str.split(' ');
//   programs.forEach(program => {
//     switch(program) {
//       case 'PUSH':
//         stack.push(register);
//         break;
//       case 'POP':
//         register = stack.pop();
//         break;
//       case 'ADD':
//         register = stack.pop() + register;
//         break;
//       case 'SUB':
//         register = register - stack.pop() ;
//         break;
//       case 'MULT':
//         register = register * stack.pop();
//         break;
//       case 'DIV':
//         register = Math.floor((register / stack.pop()));
//         break;
//       case 'REMAINDER':
//         register = register % stack.pop();
//         break;
//       case 'PRINT':
//         console.log(register);
//         break;
//       default:
//         register = Number(program);
//         break;
//     }
//   });
// }

// console.log(minilang('PRINT'));
// // 0

// console.log(minilang('5 PUSH 3 MULT PRINT'));
// // 15

// console.log(minilang('5 PRINT PUSH 3 PRINT ADD PRINT'));
// // 5
// // 3
// // 8

// console.log(minilang('5 PUSH POP PRINT'));
// // 5

// console.log(minilang('3 PUSH 4 PUSH 5 PUSH PRINT ADD PRINT POP PRINT ADD PRINT'));
// // 5
// // 10
// // 4
// // 7

// console.log(minilang('3 PUSH PUSH 7 DIV MULT PRINT'));
// // 6

// console.log(minilang('4 PUSH PUSH 7 REMAINDER MULT PRINT'));
// // 12

// console.log(minilang('-3 PUSH 5 SUB PRINT'));
// // 8

// console.log(minilang('6 PUSH'));
// // (nothing is printed because the `program` argument has no `PRINT` commands)

// P convert every number word in a string to number characters
      // input string
      // output : string
// E: five to 5 three to 3
      // assume that the case of the word doesn't matter, FIVE === 5
// D: array for iteration through each word & for the constant number words
// A: 1. create constant for number words where index correspond to the word
    // iterate through every word in the const
    // replace the word with the index of the word
    // 5. join the returned array from map with ' '
    // 6. return the joined string

// const NUMBERS = ['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'];

// function wordToDigit(str) {
//   NUMBERS.forEach((number, index) => {
//     let pattern = new RegExp(`\\b${number}\\b`, 'gi');
//     str = str.replace(pattern, index);
//   });
//   return str
// }

// console.log(wordToDigit('Please call me at five five five one two three four. Thanks. weight'))

// P: calculate the nth fibonacci number input is n output is the number
// E: 1 1 2(1+1) 3(1+2) 5(2+3) 7(3+4)
// A: f(n) = f(n-1) +  f(n-2) f1 = 1 f2 = 2
let results = {};
function fibonacci(n) {
  let result;
  if (n === 1) {
    result = 1;
  } else if (n === 2) {
    result = 1;
  } else {
    result = (results[n-1] || fibonacci(n - 1)) + (results[n-2] || fibonacci(n - 2)); 
  }
  results[n] = result;
  return result;
}

// function fibonacci(n) {
//   if (n <= 2) {
//     return 1
//   } else {
//     let f1 = 1;
//     let f2 = 1;
//     while (n - 2 > 0) {
//       let f3 = f1 + f2;
//       [f1, f2] = [f2, f3];
//       n -= 1;
//     }
//     return f2
//   }
// }

console.log(fibonacci(1));       // 1
console.log(fibonacci(2));       // 1
console.log(fibonacci(3));       // 2
console.log(fibonacci(4));       // 3
console.log(fibonacci(5));       // 5
console.log(fibonacci(12));      // 144
console.log(fibonacci(20));      // 6765